crates/driver/src/domain/competition/solution/trade.rs

```diff
@@ -23,14 +23,164 @@ pub struct Fulfillment {
+                                let max_volume_fee = executed_sell_amount_with_surplus_fee
+                                    * (eth::U256::from_f64_lossy(max_volume_factor * 100.))
+                                    / 100;
+                                // take the smaller of the two
+                                std::cmp::min(price_improvement_fee, max_volume_fee)
+                            }
+                            order::Side::Sell => {
+                                // How much `buy_token` we get for `executed` amount of `sell_token`
+                                let executed_buy_amount = executed
+                                    .0
+                                    .checked_mul(uniform_sell_price)
+                                    .ok_or(InvalidFullfilment)?
+                                    .checked_div(uniform_buy_price)
+                                    .ok_or(InvalidFullfilment)?;
+                                let executed_sell_amount = executed
+                                    .0
+                                    .checked_add(surplus_fee)
+                                    .ok_or(InvalidFullfilment)?;
+                                // What is the minimum amount of `buy_token` we have to buy based on
+                                // limit price?
+                                // Equal to full buy amount for FOK orders, otherwise scalled with
+                                // executed amount for partially fillable orders
+                                let limit_buy_amount = order
+                                    .buy
+                                    .amount
+                                    .0
+                                    .checked_mul(executed_sell_amount)
+                                    .ok_or(InvalidFullfilment)?
+                                    .checked_div(order.sell.amount.0)
+                                    .ok_or(InvalidFullfilment)?;
+                                // Bought exactly `executed_buy_amount` while the limit price is
+                                // `limit_buy_amount` Take protocol fee from the surplus
+                                let surplus = executed_buy_amount
+                                    .checked_sub(limit_buy_amount)
+                                    .unwrap_or(eth::U256::zero());
+                                let surplus_in_sell_token = surplus
+                                    .checked_mul(uniform_buy_price)
+                                    .ok_or(InvalidFullfilment)?
+                                    .checked_div(uniform_sell_price)
+                                    .ok_or(InvalidFullfilment)?;
+                                let price_improvement_fee = surplus_in_sell_token
+                                    * (eth::U256::from_f64_lossy(factor * 100.))
+                                    / 100;
+                                let max_volume_fee = executed_sell_amount
```

Comment:
We should be able to reuse this logic for volume based fee policies, right? Can we have `volume_fee` method that can be reused across arms?

Reply:
refactored.